bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableIO.java

```diff
@@ -0,0 +1,1039 @@
+          logExceptions(context, exception);
+          throw exception;
+        }
+      };
+    }
+
+  }
+
+  /**
+   * A {@link DoFn} that can write either a bounded or unbounded {@link PCollection} of
+   * {@link Mutation}s to a table specified via a {@link CloudBigtableTableConfiguration} using the
+   * BufferedMutator.
+   */
+  public static class CloudBigtableSingleTableBufferedWriteFn
+      extends BufferedMutatorDoFn<Mutation> {
+    private static final long serialVersionUID = 2L;
+    private transient BufferedMutator mutator;
+    private final String tableName;
+
+    public CloudBigtableSingleTableBufferedWriteFn(CloudBigtableTableConfiguration config) {
+      super(config);
+      tableName = config.getTableId();
+    }
+
+    @StartBundle
+    public synchronized void getBufferedMutator(StartBundleContext context)
+        throws IOException {
+      mutator = createBufferedMutator(context, tableName);
+    }
+
+    /**
+     * Performs an asynchronous mutation via {@link BufferedMutator#mutate(Mutation)}.
+     */
+    @ProcessElement
+    public void processElement(ProcessContext context) throws Exception {
+      Mutation mutation = context.element();
+      if (DOFN_LOG.isTraceEnabled()) {
+        DOFN_LOG.trace("Persisting {}", Bytes.toStringBinary(mutation.getRow()));
+      }
+      mutator.mutate(mutation);
+      mutationsCounter.inc();
+    }
+
+    /**
+     * Closes the {@link BufferedMutator} and {@link Connection}.
+     */
+    @FinishBundle
+    public synchronized void finishBundle(FinishBundleContext context) throws Exception {
+      try {
+        if (mutator != null) {
+          mutator.close();
+          mutator = null;
+        }
+      } catch (RetriesExhaustedWithDetailsException exception) {
+        exceptionsCounter.inc(exception.getCauses().size());
+        logExceptions(null, exception);
+        rethrowException(exception);
+      }
+    }
+  }
+
+  /**
+   * A {@link DoFn} that can write either a bounded or unbounded {@link PCollection} of {@link KV}
+   * of (String tableName, List of {@link Mutation}s) to the specified table.
+   *
+   * <p>NOTE: This {@link DoFn} will write {@link Put}s and {@link Delete}s, not {@link
+   * org.apache.hadoop.hbase.client.Append}s and {@link org.apache.hadoop.hbase.client.Increment}s.
+   * This limitation exists because if the batch fails partway through, Appends/Increments might be
+   * re-run, causing the {@link Mutation} to be executed twice, which is never the user's intent.
+   * Re-running a Delete will not cause any differences.  Re-running a Put isn't normally a problem,
+   * but might cause problems in some cases when the number of versions supported by the column
+   * family is greater than one.  In a case where multiple versions could be a problem, it's best to
+   * add a timestamp to the {@link Put}.
+   */
+  public static class CloudBigtableMultiTableWriteFn
+      extends BufferedMutatorDoFn<KV<String, Iterable<Mutation>>> {
```

Comment:
Can avoid carrying this class through to beam? Instead suggest users to use their own batching function?